File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Trpak/Decoders/TrinityMaterial.cs
using MiniToolbox.Trpak.Flatbuffers.TR.Model;
using MiniToolbox.Core.Utils;

namespace MiniToolbox.Trpak.Decoders
{
    /// <summary>
    /// Data-only material decoded from TRMTR / Gfx2 material files.
    /// Ported from gftool Material.cs — all GL rendering code removed.
    /// </summary>
    public class TrinityMaterial
    {
        public string Name { get; set; }
        public string ShaderName { get; set; }
        public List<TextureRef> Textures { get; } = new();
        public List<(string Name, string Value)> ShaderParams { get; } = new();
        public TRFloatParameter[] FloatParams { get; set; } = Array.Empty<TRFloatParameter>();
        public TRVec2fParameter[] Vec2Params { get; set; } = Array.Empty<TRVec2fParameter>();
        public TRVec3fParameter[] Vec3Params { get; set; } = Array.Empty<TRVec3fParameter>();
        public TRVec4fParameter[] Vec4Params { get; set; } = Array.Empty<TRVec4fParameter>();
        public TRSampler[] Samplers { get; set; } = Array.Empty<TRSampler>();

        /// <summary>
        /// Lightweight texture reference (no GL texture loading).
        /// </summary>
        public class TextureRef
        {
            public string Name { get; set; } = string.Empty;
            public string FilePath { get; set; } = string.Empty;
            public uint Slot { get; set; }
        }

        public TrinityMaterial(PathString modelPath, TRMaterial trmat)
        {
            Name = trmat.Name;
            ShaderName = ResolveShaderName(trmat.Shader?.Length > 0 ? trmat.Shader[0].Name : string.Empty);

            FloatParams = trmat.FloatParams ?? Array.Empty<TRFloatParameter>();
            Vec2Params = trmat.Vec2fParams ?? Array.Empty<TRVec2fParameter>();
            Vec3Params = trmat.Vec3fParams ?? Array.Empty<TRVec3fParameter>();
            Vec4Params = trmat.Vec4fParams ?? Array.Empty<TRVec4fParameter>();
            Samplers = trmat.Samplers ?? Array.Empty<TRSampler>();

            if (trmat.Shader != null && trmat.Shader.Length > 0 && trmat.Shader[0].Values != null)
            {
                foreach (var param in trmat.Shader[0].Values)
                    ShaderParams.Add((param.Name, param.Value));
            }

            foreach (var tex in trmat.Textures ?? Array.Empty<TRTexture>())
            {
                Textures.Add(new TextureRef
                {
                    Name = tex.Name,
                    FilePath = modelPath.Combine(tex.File),
                    Slot = tex.Slot
                });
            }
        }

        private static string ResolveShaderName(string name)
        {
            if (string.IsNullOrEmpty(name))
                return "Standard";

            return name switch
            {
                "Opaque" => "Standard",
                "Transparent" => "Transparent",
                "Hair" => "Hair",
                "SSS" => "SSS",
                "EyeClearCoat" => "EyeClearCoat",
                "Unlit" => "Unlit",
                _ => name
            };
        }
    }
}

File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Api/src/decoders/TrinityMaterial.ts
import type { PathString } from '../utils/PathString.js';
import type { TRMaterial, TRTexture, TRFloatParameter, TRVec2fParameter, TRVec3fParameter, TRVec4fParameter, TRSampler } from '../flatbuffers/TR/Model/index.js';

/**
 * Lightweight texture reference (no GL texture loading).
 */
export class TextureRef {
    public Name: string = '';
    public FilePath: string = '';
    public Slot: number = 0;
}

/**
 * Data-only material decoded from TRMTR / Gfx2 material files.
 * Ported from gftool Material.cs — all GL rendering code removed.
 */
export class TrinityMaterial {
    public Name: string = '';
    public ShaderName: string = '';
    public Textures: TextureRef[] = [];
    public ShaderParams: Array<{ Name: string; Value: string }> = [];
    public FloatParams: TRFloatParameter[] = [];
    public Vec2Params: TRVec2fParameter[] = [];
    public Vec3Params: TRVec3fParameter[] = [];
    public Vec4Params: TRVec4fParameter[] = [];
    public Samplers: TRSampler[] = [];

    constructor(modelPath: PathString, trmat: TRMaterial) {
        this.Name = trmat.Name ?? '';
        this.ShaderName = TrinityMaterial.ResolveShaderName(
            trmat.Shader?.length > 0 ? trmat.Shader[0].Name : ''
        );

        this.FloatParams = trmat.FloatParams ?? [];
        this.Vec2Params = trmat.Vec2fParams ?? [];
        this.Vec3Params = trmat.Vec3fParams ?? [];
        this.Vec4Params = trmat.Vec4fParams ?? [];
        this.Samplers = trmat.Samplers ?? [];

        if (trmat.Shader != null && trmat.Shader.length > 0 && trmat.Shader[0].Values != null) {
            for (const param of trmat.Shader[0].Values) {
                this.ShaderParams.push({ Name: param.Name, Value: param.Value });
            }
        }

        for (const tex of trmat.Textures ?? []) {
            this.Textures.push({
                Name: tex.Name,
                FilePath: modelPath.combine(tex.File),
                Slot: tex.Slot
            });
        }
    }

    private static ResolveShaderName(name: string): string {
        if (!name || name.length === 0) {
            return 'Standard';
        }

        switch (name) {
            case 'Opaque': return 'Standard';
            case 'Transparent': return 'Transparent';
            case 'Hair': return 'Hair';
            case 'SSS': return 'SSS';
            case 'EyeClearCoat': return 'EyeClearCoat';
            case 'Unlit': return 'Unlit';
            default: return name;
        }
    }
}
